File: packages/sdk/src/scorers/rule-based.ts
/**
 * Rule-Based Scorers
 *
 * Deterministic scorers that don't require LLM calls.
 */

import type { SpanWithChildren } from "@neon/shared";
import { defineScorer, type Scorer, type EvalContext } from "./base.js";

/**
 * Rule-based scorer configuration
 */
export interface RuleBasedConfig {
  /** Check function that returns a score (0-1) or boolean */
  check: (context: EvalContext) => boolean | number;
  /** Optional threshold for boolean to score conversion */
  threshold?: number;
  /** Name for the scorer */
  name?: string;
  /** Description of what the scorer checks */
  description?: string;
}

/**
 * Create a rule-based scorer
 *
 * @example
 * ```typescript
 * const toolUsedScorer = ruleBasedScorer({
 *   check: (ctx) => ctx.trace.spans.some(s => s.spanType === 'tool'),
 *   name: 'tool_used',
 * });
 * ```
 */
export function ruleBasedScorer(config: RuleBasedConfig): Scorer {
  return defineScorer({
    name: config.name ?? "rule_based",
    description: config.description,
    dataType: "numeric",
    evaluate: (context: EvalContext) => {
      const result = config.check(context);
      const value = typeof result === "boolean" ? (result ? 1 : 0) : result;

      return {
        value: Math.min(1, Math.max(0, value)),
        reason: `Rule check returned ${value}`,
      };
    },
  });
}

// ==================== Built-in Rule-Based Scorers ====================

/**
 * Check if expected tools were called
 */
export function toolSelectionScorer(expectedTools?: string[]): Scorer {
  return ruleBasedScorer({
    name: "tool_selection",
    description: "Checks if the expected tools were called",
    check: (context) => {
      const actualTools = context.trace.spans
        .filter((s: SpanWithChildren) => s.spanType === "tool")
        .map((s: SpanWithChildren) => s.toolName);

      const expected = expectedTools || (context.expected?.toolCalls as string[]) || [];

      if (expected.length === 0) {
        return actualTools.length > 0 ? 1 : 0.5;
      }

      const matches = expected.filter((t) => actualTools.includes(t));
      return matches.length / expected.length;
    },
  });
}

/**
 * Configuration for contains scorer
 */
export interface ContainsConfig {
  /** Strings to check for (case-insensitive by default) */
  expected?: string | string[];
  /** Whether to use case-sensitive matching (default: false) */
  caseSensitive?: boolean;
  /** Require all strings to match (AND) vs any string (OR). Default: true (AND) */
  matchAll?: boolean;
}

/**
 * Check if output contains expected string(s)
 *
 * @example
 * ```typescript
 * // Simple usage
 * contains('hello')
 * contains(['hello', 'world'])
 *
 * // With options
 * contains({ expected: ['error', 'warning'], matchAll: false })
 * contains({ expected: 'SUCCESS', caseSensitive: true })
 * ```
 */
export function contains(config?: string | string[] | ContainsConfig): Scorer {
  // Normalize config
  const normalizedConfig: ContainsConfig =
    typeof config === "string"
      ? { expected: [config] }
      : Array.isArray(config)
        ? { expected: config }
        : config ?? {};

  const { expected, caseSensitive = false, matchAll = true } = normalizedConfig;

  return defineScorer({
    name: "contains",
    description: "Checks if output contains expected string(s)",
    dataType: "numeric",
    evaluate: (context: EvalContext) => {
      const output = getLastOutput(context);

      // Get expected strings from config or context
      const rawExpected = expected ?? (context.expected?.outputContains as string | string[]);
      const strings = normalizeToArray(rawExpected);

      // Handle edge cases
      if (strings.length === 0) {
        return { value: 1, reason: "No expected strings specified" };
      }

      if (output === "") {
        return { value: 0, reason: "Output is empty" };
      }

      const normalizedOutput = caseSensitive ? output : output.toLowerCase();
      const matches = strings.filter((s) => {
        if (s === null || s === undefined) return false;
        const normalizedSearch = caseSensitive ? String(s) : String(s).toLowerCase();
        return normalizedOutput.includes(normalizedSearch);
      });

      const matchCount = matches.length;
      const total = strings.length;

      if (matchAll) {
        // AND mode: score is ratio of matches
        const value = matchCount / total;
        const reason =
          value === 1
            ? `All ${total} expected string(s) found`
            : `Found ${matchCount}/${total} expected strings: missing [${strings.filter((s) => !matches.includes(s)).join(", ")}]`;
        return { value, reason };
      } else {
        // OR mode: 1 if any match, 0 otherwise
        const value = matchCount > 0 ? 1 : 0;
        const reason =
          value === 1
            ? `Found matching string: "${matches[0]}"`
            : `None of the expected strings found: [${strings.join(", ")}]`;
        return { value, reason };
      }
    },
  });
}

/**
 * @deprecated Use `contains()` instead for cleaner API
 */
export function containsScorer(expected?: string[]): Scorer {
  return contains(expected);
}

/**
 * Configuration for exact match scorer
 */
export interface ExactMatchConfig {
  /** Expected output string */
  expected?: string;
  /** Whether to trim whitespace before comparison (default: true) */
  trim?: boolean;
  /** Whether to normalize whitespace (collapse multiple spaces/newlines) (default: false) */
  normalizeWhitespace?: boolean;
  /** Whether to use case-sensitive matching (default: true) */
  caseSensitive?: boolean;
}

/**
 * Check for exact output match
 *
 * @example
 * ```typescript
 * // Simple usage - expects exact string
 * exactMatch('hello world')
 *
 * // With options
 * exactMatch({ expected: 'Hello World', caseSensitive: false })
 * exactMatch({ expected: 'result', normalizeWhitespace: true })
 * ```
 */
export function exactMatch(config?: string | ExactMatchConfig): Scorer {
  // Normalize config
  const normalizedConfig: ExactMatchConfig =
    typeof config === "string" ? { expected: config } : config ?? {};

  const {
    expected,
    trim = true,
    normalizeWhitespace = false,
    caseSensitive = true,
  } = normalizedConfig;

  return defineScorer({
    name: "exact_match",
    description: "Checks for exact output match",
    dataType: "numeric",
    evaluate: (context: EvalContext) => {
      const rawOutput = getLastOutput(context);
      const rawExpected = expected ?? (context.expected?.output as string);

      // Handle null/undefined expected
      if (rawExpected === null || rawExpected === undefined) {
        return { value: 1, reason: "No expected output specified" };
      }

      // Handle null/undefined output
      if (rawOutput === null || rawOutput === undefined) {
        return { value: 0, reason: "Output is null or undefined" };
      }

      // Normalize strings
      let output = String(rawOutput);
      let expectedStr = String(rawExpected);

      if (trim) {
        output = output.trim();
        expectedStr = expectedStr.trim();
      }

      if (normalizeWhitespace) {
        output = output.replace(/\s+/g, " ");
        expectedStr = expectedStr.replace(/\s+/g, " ");
      }

      if (!caseSensitive) {
        output = output.toLowerCase();
        expectedStr = expectedStr.toLowerCase();
      }

      const matches = output === expectedStr;

      if (matches) {
        return { value: 1, reason: "Output matches expected exactly" };
      }

      // Provide helpful diff info for debugging
      const outputPreview = output.length > 50 ? `${output.slice(0, 50)}...` : output;
      const expectedPreview =
        expectedStr.length > 50 ? `${expectedStr.slice(0, 50)}...` : expectedStr;

      return {
        value: 0,
        reason: `Output "${outputPreview}" does not match expected "${expectedPreview}"`,
      };
    },
  });
}

/**
 * @deprecated Use `exactMatch()` instead for cleaner API
 */
export function exactMatchScorer(expected?: string): Scorer {
  return exactMatch(expected);
}

/**
 * Check if output matches JSON structure
 */
export function jsonMatchScorer(expected?: object): Scorer {
  return ruleBasedScorer({
    name: "json_match",
    description: "Checks if output matches expected JSON structure",
    check: (context) => {
      const output = getLastOutput(context);
      const expectedObj = expected || context.expected;

      try {
        const parsed = JSON.parse(output);
        return deepMatch(parsed, expectedObj) ? 1 : 0;
      } catch {
        return 0;
      }
    },
  });
}

/**
 * Score based on latency
 */
export function latencyScorer(thresholds?: {
  excellent?: number;
  good?: number;
  acceptable?: number;
}): Scorer {
  const { excellent = 1000, good = 5000, acceptable = 10000 } =
    thresholds || {};

  return ruleBasedScorer({
    name: "latency",
    description: "Scores based on execution latency",
    check: (context) => {
      const duration = context.trace.trace.durationMs;

      if (duration <= excellent) return 1.0;
      if (duration <= good) return 0.8;
      if (duration <= acceptable) return 0.6;
      return 0.4;
    },
  });
}

/**
 * Score based on error rate
 */
export function errorRateScorer(): Scorer {
  return ruleBasedScorer({
    name: "error_rate",
    description: "Scores based on span error rate",
    check: (context) => {
      const spans = context.trace.spans;
      if (spans.length === 0) return 1;

      const errors = spans.filter((s: SpanWithChildren) => s.status === "error").length;
      return 1 - errors / spans.length;
    },
  });
}

/**
 * Score based on token efficiency
 */
export function tokenEfficiencyScorer(thresholds?: {
  excellent?: number;
  good?: number;
  acceptable?: number;
}): Scorer {
  const { excellent = 1000, good = 5000, acceptable = 10000 } =
    thresholds || {};

  return ruleBasedScorer({
    name: "token_efficiency",
    description: "Scores based on total token usage",
    check: (context) => {
      const totalTokens = context.trace.spans
        .filter((s: SpanWithChildren) => s.spanType === "generation")
        .reduce((sum: number, s: SpanWithChildren) => sum + (s.totalTokens || 0), 0);

      if (totalTokens <= excellent) return 1.0;
      if (totalTokens <= good) return 0.8;
      if (totalTokens <= acceptable) return 0.6;
      return 0.4;
    },
  });
}

/**
 * Check if trace completed successfully
 */
export function successScorer(): Scorer {
  return ruleBasedScorer({
    name: "success",
    description: "Checks if trace completed successfully",
    check: (context) => {
      return context.trace.trace.status === "ok" ? 1 : 0;
    },
  });
}

/**
 * Score based on iteration count
 */
export function iterationScorer(maxIterations: number = 10): Scorer {
  return ruleBasedScorer({
    name: "iterations",
    description: "Scores based on number of iterations",
    check: (context) => {
      // Count generation spans as proxy for iterations
      const iterations = context.trace.spans.filter(
        (s: SpanWithChildren) => s.spanType === "generation"
      ).length;

      if (iterations <= 1) return 1.0;
      if (iterations <= 3) return 0.9;
      if (iterations <= 5) return 0.7;
      if (iterations <= maxIterations) return 0.5;
      return 0.3;
    },
  });
}

// ==================== Utility Functions ====================

/**
 * Get the last output from the trace
 */
function getLastOutput(context: EvalContext): string {
  const generations = context.trace.spans.filter(
    (s: SpanWithChildren) => s.spanType === "generation"
  );
  const lastGen = generations[generations.length - 1];
  return lastGen?.output || "";
}

/**
 * Normalize a value to an array
 */
function normalizeToArray<T>(value: T | T[] | undefined | null): T[] {
  if (value === null || value === undefined) return [];
  return Array.isArray(value) ? value : [value];
}

/**
 * Deep match two objects
 */
function deepMatch(actual: unknown, expected: unknown): boolean {
  if (expected === undefined) return true;
  if (typeof expected !== typeof actual) return false;

  if (typeof expected === "object" && expected !== null) {
    if (Array.isArray(expected)) {
      if (!Array.isArray(actual)) return false;
      return expected.every((item, i) => deepMatch(actual[i], item));
    }

    for (const key of Object.keys(expected)) {
      if (
        !deepMatch(
          (actual as Record<string, unknown>)[key],
          (expected as Record<string, unknown>)[key]
        )
      ) {
        return false;
      }
    }
    return true;
  }

  return actual === expected;
}

File: packages/neon-sdk-python/neon_sdk/scorers/rule_based.py
"""
Rule-Based Scorers

Deterministic scorers that don't require LLM calls.
"""

from __future__ import annotations

import json
import re
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from neon_sdk.types import ScoreDataType, SpanType, SpanWithChildren

from .base import EvalContext, ScorerConfig, ScoreResult, ScorerImpl, define_scorer

# =============================================================================
# Rule-Based Scorer Factory
# =============================================================================


@dataclass
class RuleBasedConfig:
    """Rule-based scorer configuration."""

    check: Callable[[EvalContext], bool | float]
    name: str = "rule_based"
    description: str | None = None
    threshold: float | None = None


def rule_based_scorer(config: RuleBasedConfig) -> ScorerImpl:
    """
    Create a rule-based scorer.

    Example:
        ```python
        tool_used_scorer = rule_based_scorer(RuleBasedConfig(
            check=lambda ctx: any(s.span_type == SpanType.TOOL for s in ctx.trace.spans),
            name='tool_used',
        ))
        ```
    """

    def evaluate(context: EvalContext) -> ScoreResult:
        result = config.check(context)
        value = 1.0 if result is True else (0.0 if result is False else float(result))
        return ScoreResult(
            value=max(0.0, min(1.0, value)),
            reason=f"Rule check returned {value}",
        )

    return define_scorer(
        ScorerConfig(
            name=config.name,
            description=config.description,
            data_type=ScoreDataType.NUMERIC,
            evaluate=evaluate,
        )
    )


# =============================================================================
# Utility Functions
# =============================================================================


def _flatten_spans(spans: list[SpanWithChildren]) -> list[SpanWithChildren]:
    """Flatten a span tree into a flat list."""
    result: list[SpanWithChildren] = []

    def traverse(span: SpanWithChildren) -> None:
        result.append(span)
        for child in span.children:
            traverse(child)

    for s in spans:
        traverse(s)
    return result


def _get_last_output(context: EvalContext) -> str:
    """Get the last output from the trace."""
    flat = _flatten_spans(context.trace.spans)
    generations = [s for s in flat if s.span_type == SpanType.GENERATION]
    if generations:
        return generations[-1].output or ""
    return ""


def _normalize_to_list(value: str | list[str] | None) -> list[str]:
    """Normalize a value to a list."""
    if value is None:
        return []
    if isinstance(value, str):
        return [value]
    return list(value)


def _deep_match(actual: Any, expected: Any) -> bool:
    """Deep match two objects."""
    if expected is None:
        return True
    if not isinstance(actual, type(expected)):
        return False

    if isinstance(expected, dict):
        if not isinstance(actual, dict):
            return False
        return all(
            _deep_match(actual.get(key), expected[key]) for key in expected
        )

    if isinstance(expected, list):
        if not isinstance(actual, list):
            return False
        return all(_deep_match(actual[i], item) for i, item in enumerate(expected))

    return bool(actual == expected)


# =============================================================================
# Built-in Rule-Based Scorers
# =============================================================================


def tool_selection_scorer(expected_tools: list[str] | None = None) -> ScorerImpl:
    """
    Check if expected tools were called.

    Example:
        ```python
        scorer = tool_selection_scorer(['search', 'calculate'])
        ```
    """

    def check(context: EvalContext) -> float:
        flat = _flatten_spans(context.trace.spans)
        actual_tools = [s.tool_name for s in flat if s.span_type == SpanType.TOOL and s.tool_name]

        expected = expected_tools
        if expected is None and context.expected:
            expected = context.expected.get("toolCalls", [])
        if not expected:
            return 1.0 if actual_tools else 0.5

        matches = [t for t in expected if t in actual_tools]
        return len(matches) / len(expected)

    return rule_based_scorer(
        RuleBasedConfig(
            name="tool_selection",
            description="Checks if the expected tools were called",
            check=check,
        )
    )


# =============================================================================
# Contains Scorer
# =============================================================================


@dataclass
class ContainsConfig:
    """Configuration for contains scorer."""

    expected: str | list[str] | None = None
    case_sensitive: bool = False
    match_all: bool = True


def contains(config: str | list[str] | ContainsConfig | None = None) -> ScorerImpl:
    """
    Check if output contains expected string(s).

    Example:
        ```python
        # Simple usage
        contains('hello')
        contains(['hello', 'world'])

        # With options
        contains(ContainsConfig(expected=['error', 'warning'], match_all=False))
        contains(ContainsConfig(expected='SUCCESS', case_sensitive=True))
        ```
    """
    # Normalize config
    if config is None:
        normalized = ContainsConfig()
    elif isinstance(config, str):
        normalized = ContainsConfig(expected=[config])
    elif isinstance(config, list):
        normalized = ContainsConfig(expected=config)
    else:
        normalized = config

    def evaluate(context: EvalContext) -> ScoreResult:
        output = _get_last_output(context)

        # Get expected strings from config or context
        raw_expected = normalized.expected
        if raw_expected is None and context.expected:
            raw_expected = context.expected.get("outputContains")
        strings = _normalize_to_list(raw_expected)

        # Handle edge cases
        if not strings:
            return ScoreResult(value=1.0, reason="No expected strings specified")
        if not output:
            return ScoreResult(value=0.0, reason="Output is empty")

        normalized_output = output if normalized.case_sensitive else output.lower()
        matches = []
        for s in strings:
            if s is None:
                continue
            normalized_search = str(s) if normalized.case_sensitive else str(s).lower()
            if normalized_search in normalized_output:
                matches.append(s)

        match_count = len(matches)
        total = len(strings)

        if normalized.match_all:
            # AND mode: score is ratio of matches
            value = match_count / total
            if value == 1.0:
                reason = f"All {total} expected string(s) found"
            else:
                missing = [s for s in strings if s not in matches]
                reason = f"Found {match_count}/{total} expected strings: missing {missing}"
            return ScoreResult(value=value, reason=reason)
        else:
            # OR mode: 1 if any match, 0 otherwise
            value = 1.0 if match_count > 0 else 0.0
            if value == 1.0:
                reason = f'Found matching string: "{matches[0]}"'
            else:
                reason = f"None of the expected strings found: {strings}"
            return ScoreResult(value=value, reason=reason)

    return define_scorer(
        ScorerConfig(
            name="contains",
            description="Checks if output contains expected string(s)",
            data_type=ScoreDataType.NUMERIC,
            evaluate=evaluate,
        )
    )


# Legacy alias
def contains_scorer(expected: list[str] | None = None) -> ScorerImpl:
    """
    Deprecated: Use `contains()` instead.
    """
    return contains(expected)


# =============================================================================
# Exact Match Scorer
# =============================================================================


@dataclass
class ExactMatchConfig:
    """Configuration for exact match scorer."""

    expected: str | None = None
    trim: bool = True
    normalize_whitespace: bool = False
    case_sensitive: bool = True


def exact_match(config: str | ExactMatchConfig | None = None) -> ScorerImpl:
    """
    Check for exact output match.

    Example:
        ```python
        # Simple usage - expects exact string
        exact_match('hello world')

        # With options
        exact_match(ExactMatchConfig(expected='Hello World', case_sensitive=False))
        exact_match(ExactMatchConfig(expected='result', normalize_whitespace=True))
        ```
    """
    # Normalize config
    if config is None:
        normalized = ExactMatchConfig()
    elif isinstance(config, str):
        normalized = ExactMatchConfig(expected=config)
    else:
        normalized = config

    def evaluate(context: EvalContext) -> ScoreResult:
        raw_output = _get_last_output(context)
        raw_expected = normalized.expected
        if raw_expected is None and context.expected:
            raw_expected = context.expected.get("output")

        # Handle null/undefined expected
        if raw_expected is None:
            return ScoreResult(value=1.0, reason="No expected output specified")

        # Handle null/undefined output
        if raw_output is None:
            return ScoreResult(value=0.0, reason="Output is null or undefined")

        # Normalize strings
        output = str(raw_output)
        expected_str = str(raw_expected)

        if normalized.trim:
            output = output.strip()
            expected_str = expected_str.strip()

        if normalized.normalize_whitespace:
            output = re.sub(r"\s+", " ", output)
            expected_str = re.sub(r"\s+", " ", expected_str)

        if not normalized.case_sensitive:
            output = output.lower()
            expected_str = expected_str.lower()

        matches = output == expected_str

        if matches:
            return ScoreResult(value=1.0, reason="Output matches expected exactly")

        # Provide helpful diff info for debugging
        output_preview = output[:50] + "..." if len(output) > 50 else output
        expected_preview = expected_str[:50] + "..." if len(expected_str) > 50 else expected_str

        return ScoreResult(
            value=0.0,
            reason=f'Output "{output_preview}" does not match expected "{expected_preview}"',
        )

    return define_scorer(
        ScorerConfig(
            name="exact_match",
            description="Checks for exact output match",
            data_type=ScoreDataType.NUMERIC,
            evaluate=evaluate,
        )
    )


# Legacy alias
def exact_match_scorer(expected: str | None = None) -> ScorerImpl:
    """
    Deprecated: Use `exact_match()` instead.
    """
    return exact_match(expected)


# =============================================================================
# JSON Match Scorer
# =============================================================================


def json_match_scorer(expected: dict[str, Any] | None = None) -> ScorerImpl:
    """
    Check if output matches JSON structure.

    Example:
        ```python
        scorer = json_match_scorer({'status': 'success', 'count': 5})
        ```
    """

    def check(context: EvalContext) -> float:
        output = _get_last_output(context)
        expected_obj = expected or context.expected

        try:
            parsed = json.loads(output)
            return 1.0 if _deep_match(parsed, expected_obj) else 0.0
        except (json.JSONDecodeError, TypeError):
            return 0.0

    return rule_based_scorer(
        RuleBasedConfig(
            name="json_match",
            description="Checks if output matches expected JSON structure",
            check=check,
        )
    )


# =============================================================================
# Latency Scorer
# =============================================================================


@dataclass
class LatencyThresholds:
    """Latency thresholds for scoring."""

    excellent: int = 1000
    good: int = 5000
    acceptable: int = 10000


def latency_scorer(thresholds: LatencyThresholds | None = None) -> ScorerImpl:
    """
    Score based on latency.

    Example:
        ```python
        scorer = latency_scorer(LatencyThresholds(excellent=500, good=2000, acceptable=5000))
        ```
    """
    t = thresholds or LatencyThresholds()

    def check(context: EvalContext) -> float:
        duration = context.trace.trace.duration_ms

        if duration <= t.excellent:
            return 1.0
        if duration <= t.good:
            return 0.8
        if duration <= t.acceptable:
            return 0.6
        return 0.4

    return rule_based_scorer(
        RuleBasedConfig(
            name="latency",
            description="Scores based on execution latency",
            check=check,
        )
    )


# =============================================================================
# Error Rate Scorer
# =============================================================================


def error_rate_scorer() -> ScorerImpl:
    """
    Score based on span error rate.

    Example:
        ```python
        scorer = error_rate_scorer()
        ```
    """

    def check(context: EvalContext) -> float:
        flat = _flatten_spans(context.trace.spans)
        if not flat:
            return 1.0

        errors = sum(1 for s in flat if s.status.value == "error")
        return 1.0 - errors / len(flat)

    return rule_based_scorer(
        RuleBasedConfig(
            name="error_rate",
            description="Scores based on span error rate",
            check=check,
        )
    )


# =============================================================================
# Token Efficiency Scorer
# =============================================================================


@dataclass
class TokenThresholds:
    """Token usage thresholds for scoring."""

    excellent: int = 1000
    good: int = 5000
    acceptable: int = 10000


def token_efficiency_scorer(thresholds: TokenThresholds | None = None) -> ScorerImpl:
    """
    Score based on token efficiency.

    Example:
        ```python
        scorer = token_efficiency_scorer(TokenThresholds(excellent=500, good=2000))
        ```
    """
    t = thresholds or TokenThresholds()

    def check(context: EvalContext) -> float:
        flat = _flatten_spans(context.trace.spans)
        generations = [s for s in flat if s.span_type == SpanType.GENERATION]
        total_tokens = sum(s.total_tokens or 0 for s in generations)

        if total_tokens <= t.excellent:
            return 1.0
        if total_tokens <= t.good:
            return 0.8
        if total_tokens <= t.acceptable:
            return 0.6
        return 0.4

    return rule_based_scorer(
        RuleBasedConfig(
            name="token_efficiency",
            description="Scores based on total token usage",
            check=check,
        )
    )


# =============================================================================
# Success Scorer
# =============================================================================


def success_scorer() -> ScorerImpl:
    """
    Check if trace completed successfully.

    Example:
        ```python
        scorer = success_scorer()
        ```
    """

    def check(context: EvalContext) -> bool:
        return context.trace.trace.status.value == "ok"

    return rule_based_scorer(
        RuleBasedConfig(
            name="success",
            description="Checks if trace completed successfully",
            check=check,
        )
    )


# =============================================================================
# Iteration Scorer
# =============================================================================


def iteration_scorer(max_iterations: int = 10) -> ScorerImpl:
    """
    Score based on iteration count.

    Example:
        ```python
        scorer = iteration_scorer(max_iterations=5)
        ```
    """

    def check(context: EvalContext) -> float:
        flat = _flatten_spans(context.trace.spans)
        iterations = sum(1 for s in flat if s.span_type == SpanType.GENERATION)

        if iterations <= 1:
            return 1.0
        if iterations <= 3:
            return 0.9
        if iterations <= 5:
            return 0.7
        if iterations <= max_iterations:
            return 0.5
        return 0.3

    return rule_based_scorer(
        RuleBasedConfig(
            name="iterations",
            description="Scores based on number of iterations",
            check=check,
        )
    )


__all__ = [
    # Factory
    "RuleBasedConfig",
    "rule_based_scorer",
    # Contains
    "ContainsConfig",
    "contains",
    "contains_scorer",
    # Exact match
    "ExactMatchConfig",
    "exact_match",
    "exact_match_scorer",
    # Other scorers
    "tool_selection_scorer",
    "json_match_scorer",
    "LatencyThresholds",
    "latency_scorer",
    "error_rate_scorer",
    "TokenThresholds",
    "token_efficiency_scorer",
    "success_scorer",
    "iteration_scorer",
]
